R/summary_funs.R

```diff
@@ -521,3 +552,38 @@ get_nfeat_baseline <- function(object, baseline, stat, ...) {
   }
   return(nfeat_baseline)
 }
+
+## Difference estimation using SRS-WOR sampling (Magnusson et al., 2020)
+## copied from loo:::srs_diff_est and added weights
+.srs_diff_est_w <- function(y_approx, y, y_idx, wobs = 1) {
+
+  N <- length(y_approx)
+  m <- length(y)
+  y_approx_m <- y_approx[y_idx]
+  wobs_m <- 1
+  if (length(wobs) > 1) {
+    wobs_m <- wobs[y_idx]
+    wobs_m <- length(wobs_m) * wobs_m / sum(wobs_m)
+    wobs <- length(wobs) * wobs / sum(wobs)
+  }
+
+  e_i <- y - y_approx_m
+  t_pi_tilde <- sum(wobs * y_approx)
+  t_pi2_tilde <- sum(wobs * y_approx^2)
+  t_e <- N * mean(wobs_m * e_i)
+  t_hat_epsilon <- N * mean(wobs_m * (y^2 - y_approx_m^2))
+
+  est_list <- nlist(m, N)
+  # eq (7)
+  est_list$y_hat <- t_pi_tilde + t_e
+  # eq (8)
+  var_e_i <- m / (m - 1) * (mean(wobs_m * e_i^2) - mean(wobs_m * e_i)^2)
```

Comment:
Was there a specific reason not to use `.weighted_sd()` and square it here? It should give the same result (see reprex below), but in a computationally slightly more stable way (see https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance). For comparison: loo has `var(e_i)` here (see https://github.com/stan-dev/loo/blob/6e7001e35bf74f9829ba9ab0618c1e860c7d7d94/R/loo_subsample.R#L1192).

Reprex:
``` r
devtools::load_all(".")
#> ℹ Loading projpred
#> This is projpred version 2.8.0.9000.
#> NOTE: In projpred 2.7.0, the default search method was set to "forward" (for all kinds of models).
m <- 20
set.seed(347)
e_i <- rnorm(m)
wobs_m <- rgamma(m, shape = 2)
wobs_m <- length(wobs_m) * wobs_m / sum(wobs_m)
( var_e_i <- m / (m - 1) * (mean(wobs_m * e_i^2) - mean(wobs_m * e_i)^2) )
#> [1] 1.242891
( var_e_i_via_weighted_sd <- .weighted_sd(e_i, w = wobs_m)^2 )
#> [1] 1.242891
```

<sup>Created on 2024-08-27 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>